### ISSUE TITLE
optimus pipeline: fastqprocess running out of memory because of whitelist?

### ISSUE THREAD

**Issue by plijnzaad** [broadinstitute/warp#283]
Dear all, 

we are trying to get the optimus pipeline running with Singularity on our local HPC. The process kept running out of memory, and it turns out that the whitelist is the culprit: the 6.8 M lines that are present in the latest version ( 3M-february-2018.txt.gz; see https://github.com/10XGenomics/cellranger/tree/master/lib/python/cellranger/barcodes ) seem to be too much for `fastqprocess`. Even 80 GB is not enough whereas the original wdl file specifies just 40 GB. 

Should we be using an older/much smaller whitelist and if so, which one  (e.g. 737K-august-2016.txt) ? 

Or is something else the matter? Any input appreciated!

Cheers,

Philip

**Comment by plijnzaad** [comment]
For completeness sake, we were using Singularity images created from this docker image:
docker://quay.io/humancellatlas/secondary-analysis-sctools@sha256:bf70bbdb166edc857875899209c620f709b4a9a2570b2cad4abb419926a258f8

**Comment by plijnzaad** [comment]
And just another datapoint: the 3M-february-2018.txt.gz file is a two-column file, each containing the exact same 6794880 unique barcodes ( but in different order). We now ran fastqprocess with just the first column and it did run to completion and with reasonable output  (~ 30% barcodes were corrected;  2% uncorrectable).  Does this make sense?

**Comment by kbergin** [comment]
Hi @plijnzaad apologies for the delay in getting back to you! I've asked my team to review your question and we'll respond here.

**Comment by kishorikonwar** [comment]
@plijnzaad  Those stats for the barcodes seem to be in the right range.

**Comment by plijnzaad** [comment]
Thanks for checking this. I just checked and there are in fact two files called 3M-february-2018.txt, one in 
https://github.com/10XGenomics/cellranger/tree/master/lib/python/cellranger/barcodes/translation  (has 2 columns, see above) and one in https://github.com/10XGenomics/cellranger/tree/master/lib/python/cellranger/barcodes (which has the exact same barcodes). I guess we used the wrong (2-column) file, with `fastqprocess` then simply adding _both_ barcodes on each line to some internal structure. It would have been helpful if `fastqprocess` either ignores the duplicates in the whitelist (if possible with warning), or simply crashes (with warning) about this error. 

If (in the future?) different whitelists _are_ needed for the P7 and the P5 side of each PE read, I think some memory optimization would be required (if this is indeed the problem), I would think that 6.5M 16-char words should easily fit in 1GB.
